Refactors weather into a subsystem

Xhuis opened the pull request:
Weather has been refactored from a weather control machine on the lavaland z-level into a subsystem. All existing weather has been changed to accommodate this change, and their code has been optimized by using `addtimer()` instead of `sleep()`.

The new subsystem also supports adding weather to any z-level; for instance, if you made a weather called `/datum/weather/rain_storm` and made it target z-level 3 with a 100% probability, the rain storm would occur as much as possible with 5-10 minute intermissions. These intermissions take into account the weather's duration.

Commit:
Weather subsystem

RemieRichards commented on code/controllers/subsystem/weather.dm:
```diff
@@ -0,0 +1,44 @@
+		existing_weather += new W
+
+/datum/subsystem/weather/fire()
+	for(var/datum/weather/W in processing)
```
this should be an istypeless forloop.

RemieRichards commented on code/controllers/subsystem/weather.dm:
```diff
@@ -0,0 +1,44 @@
+				W.impact(L)
+	for(var/Z in eligible_zlevels)
+		var/list/possible_weather_for_this_z = list()
+		for(var/datum/weather/WE in existing_weather)
```
this should be an istypeless forloop.

RemieRichards commented on code/controllers/subsystem/weather.dm:
```diff
@@ -0,0 +1,44 @@
+/datum/subsystem/weather/proc/run_weather(weather_name)
+	if(!weather_name)
+		return
+	for(var/datum/weather/W in existing_weather)
```
this should be an istypeless forloop.

RemieRichards commented on code/datums/weather/weather.dm:
```diff
@@ -0,0 +1,120 @@
+	if(stage == STARTUP_STAGE)
+		return
+	stage = STARTUP_STAGE
+	for(var/area/A in get_areas(area_type))
```
this should be an istypeless forloop.

RemieRichards commented on code/datums/weather/weather.dm:
```diff
@@ -0,0 +1,120 @@
+			impacted_areas |= A
+	weather_duration = rand(weather_duration_lower, weather_duration_upper)
+	update_areas()
+	for(var/mob/M in player_list)
```
this should be an istypeless forloop.

RemieRichards commented on code/datums/weather/weather.dm:
```diff
@@ -0,0 +1,120 @@
+		return
+	stage = MAIN_STAGE
+	update_areas()
+	for(var/mob/M in player_list)
```
this should be an istypeless forloop.

RemieRichards commented on code/datums/weather/weather.dm:
```diff
@@ -0,0 +1,120 @@
+		return
+	stage = WIND_DOWN_STAGE
+	update_areas()
+	for(var/mob/M in player_list)
```
this should be an istypeless forloop.

RemieRichards commented on code/datums/weather/weather.dm:
```diff
@@ -0,0 +1,120 @@
+	return
+
+/datum/weather/proc/update_areas()
+	for(var/area/N in impacted_areas)
```
this should be an istypeless forloop.

RemieRichards commented on code/datums/weather/weather_types.dm:
```diff
@@ -0,0 +1,112 @@
+	target_z = ZLEVEL_STATION
+
+/datum/weather/advanced_darkness/update_areas()
+	for(var/area/A in impacted_areas)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm:
```diff
@@ -118,37 +118,37 @@
-				linked_machine = controller
-				break
+	var/datum/weather/ash_storm/A
+	for(var/datum/weather/W in SSweather.existing_weather)
```
this should be an istypeless forloop.

optimumtact commented on code/modules/ruins/lavaland_ruin_code.dm:
```diff
@@ -1,36 +1,5 @@
-		return QDEL_HINT_LETMELIVE
-
-
-/obj/structure/fans/tiny/invisible //For blocking air in ruin doorways
```
Did you mean to remove this?

coiax replied:
TINY FAN

Commit:
Removes the weather controler + readds tiny fan

Xhuis commented:
Tiny fan re-added and weather controller removed from lavaland. Didn't use the map merger because it seems to have issues with `lavaland.dmm`.

optimumtact commented:
what kind of issues

Xhuis commented:
You can see it in the other commit. It replaces a single tile that has two medkits _every time_ and it doesn't seem to save removals.

coiax commented:
@xxalpha @xxalpha @xxalpha 

Say their name three times to summon them.

xxalpha commented:
Mmmiracles edited lavaland just fine.

Xhuis commented:
Guess it's just me, then. The TGM mapmerger refuses to work for me with lavaland.

Xhuis commented:
Closing this until I can figure out why the mapmerger won't work for me.

Commit:
Conflicts 1

Commit:
Conflicts 2

Xhuis commented:
It seems like other people are having the same issue. For the time being, I'm just going to remove the functionality of the weather controller. New commit coming soon.

Commit:
Implements changes based on feedback

Xhuis commented:
* Replaced a bunch of for loops with istypeless for loops.
* Removed map edits and removed the functionality of the weather machine pending its removal.
* Re-added invisible tiny fans.

Commit:
Removed an unintended change

Xhuis commented:
This is no longer WIP and should be merge-ready.

Commit:
Fixes a gamebreaking bug

KorPhaeron commented on code/modules/admin/secrets.dm:
```diff
@@ -427,8 +427,7 @@
 		if("floorlava")
-			var/datum/weather/floor_is_lava/storm = new /datum/weather/floor_is_lava
-			storm.weather_start_up()
+			SSweather.run_weather("the floor is lava")
```
why would you merge these magic string memes @optimumtact

KorPhaeron replied:
@JJRcop

Xhuis replied:
Should I be using defines? An ID var in weather itself?

xxalpha replied:
Yes, number values with descriptive defines. There is no reason to use strings for this sort of stuff.

KorPhaeron commented on code/controllers/subsystem/weather.dm:
```diff
@@ -0,0 +1,50 @@
+		var/datum/weather/W = V
+		existing_weather |= new W
+
+/datum/subsystem/weather/proc/run_weather(weather_name)
```
@JJRcop look at this aaaaaaaa
